Pipe gas visuals

ninjanomnom opened the pull request:
## About The Pull Request

Primarily this pr is all about getting pipes to have a visual display for the gas within them. A couple other things of note have been done to make that easier though:

- A sprite generator that outputs to a dmi has been made for all pipe variants and the layers. This is because I didn't want to work with hundreds of pipe sprites just to do minor changes. In the future I would like to generate things like this during a github action but for the moment this is just a helper you can use for generating the pipe dmi files.
- Some minor reorganization of the code in the pipe dm file.
- Some doc comments on things that bothered me

## Why It's Good For The Game

Gives a bit of visual feedback when working with pipes that should hopefully make people more comfortable around them if they can easily see if the pipes are empty or are filled with an unknown gas mixture. Can also serve as a warning to those keeping an eye out for things like plasma floods.

:cl: ninjanomnom
add: Pipes now have a colored visual display that shows their contents at a glance.
/:cl:

~~The colors for gases could stand to be better, if anyone wants to suggest alternatives to what I've used please do as I put only a bare minimum of thought into each choice.~~ I've switched to using the gas colors defined in tgui constants.

![dreamseeker_2023-09-26_07-32-56](https://github.com/tgstation/tgstation/assets/1234602/0a3a5636-c9b9-4046-83b9-0d116f8232d1)

ninjanomnom committed:
Automatic generation of pipe sprites from some junctions and straights

ninjanomnom committed:
Basic gas overlays

ninjanomnom committed:
Functional but the cross version is broken and commented out

ninjanomnom committed:
Rewrite of the generator to be as sequential as reasonable

Finally dealt with the last of the bad icon operation runtimes

ninjanomnom committed:
Makes pipelines use the correct proc to assign itself as parent

ninjanomnom committed:
Adds a preliminary primary color to gas types

These were mostly chosen at random from previously existing color choices for gases, they should probably have more thought put into the choices.

ninjanomnom committed:
Basic functional pipe gas visuals

I really don't like how set_pipenet and replace_pipenet work but dealing with that mess is outside of the scope of this feature.

ninjanomnom committed:
Slight visual improvement to gas visuals

ninjanomnom committed:
Cleans up the graphics a bit and makes only normal pipes have a visual

ninjanomnom committed:
Removes the no longer used signal

ninjanomnom committed:
Removes missed unused var from older code

Wallemations commented:
You reminded me how we need new pipes but that would entail like 2000 sprites


Good idea though I love visual feedback

Serxule commented:
could these be optional pipes? its a nice qol but itd also make plasmafloods basically useless unless you prep like every air alarm

distributivgesetz commented:
are the colors supposed to indicate the color of the gas?

Ghilker commented on code/modules/atmospherics/machinery/pipes/pipes.dm:
```diff
@@ -1,15 +1,22 @@
 	paintable = TRUE
 
+	/// Determines if this pipe will be given gas visuals
+	var/has_gas_visuals = TRUE
```
this should be opt-in instead.

ninjanomnom replied:
Except that this isn't an abstract type and can be used by itself with gas visuals

LemonInTheDark replied:
tfw your leaf is also a branch

Ghilker replied:
But you won't see this object in game, so it is an "abstract" type.. Kind of

DATA-xPUNGED commented:
> You reminded me how we need new pipes but that would entail like 2000 sprites

Ideally someone would rework how pipe sprites work in general, but good luck to whoever does that

intercepti0n commented:
I agree with @Serxule , would be nice to have a window on a NEW TYPE of pipe to see what gas is inside, but not on default one. It's messy and meter tells a lot more.

intercepti0n commented on code/modules/atmospherics/machinery/datum_pipeline.dm:
```diff
@@ -42,6 +49,13 @@
+	if(new_air == air)
+		return
+	air = new_air
+	CalculateGasmixColor(air)
```
Why don't you use same method of generateing gas overlay as a gas can?

ninjanomnom replied:
Much much more expensive and the window is small enough that the gas wouldn't display well

Jacquerel commented:
> I agree with @Serxule , would be nice to have a window on a NEW TYPE of pipe to see what gas is inside, but not on default one. It's messy and meter tells a lot more.

It's extremely preferable for this to be the default. We want game information to be visible to players' eyes if possible, not siloed into a text box which is only displayed when using a device.

BlueMemesauce commented:
it doesn't make sense since most gases like oxygen, carbon dioxide, and nitrogen are colorless so you wouldn't be able to see anything through a pipe window

ninjanomnom commented:
It's not meant to be a window but more a digital display that reacts to detected gases within, at least that's the fluff explanation in my head.

SimplySmiley123 commented:
> > I agree with @Serxule , would be nice to have a window on a NEW TYPE of pipe to see what gas is inside, but not on default one. It's messy and meter tells a lot more.
> 
> It's extremely preferable for this to be the default. We want game information to be visible to players' eyes if possible, not siloed into a text box which is only displayed when using a device.

What imo would be the best compromise is for this to be a new pipe type used 2-3 times in few places atmos techs would be checking or wherever it makes sense for a pipe to have a check for if any gasses are flowing through it.
Cuz I don't think every single pipe should have this everywhere across the whole station.

ninjanomnom committed:
Changes colors to match TGUI used colors for gases

LemonInTheDark commented:
at base I'm down with this. I am a bit worried about the plasmaflood thing but I think it'll be ok enough since pipes are typically out of the way. we'll see

TM'ing on manuel to see if the maptick is alright, will need to be up for a while

ninjanomnom committed:
Fixes a color adapter invalid overlay icon state

The issue existed previously but was just getting a blank icon instead of the error icon I added

rdragan commented:
Plasma flood worries aside:

- Subjective but this looks ugly, like debug build coder sprite ugly
- We already have gas meters which look much better and also work better at giving at-a-glance information
- Avoiding pipe clutter in atmos is almost an art in itself, this makes even simple pipe setups look visually heavy and cluttered.
- Atmos already has major lag issues once pop goes over ~60 players and the atmos tech job becomes literally useless on highpop terry (consistently hitting 90 players every evening) as atmos crawls to a halt with 2 ticks per minute and even SM delams becoming impossible. Why are you adding additional load on a system that severely needs looking into more optimization?

> TM'ing on manuel to see if the maptick is alright, will need to be up for a while
@LemonInTheDark 

Can you please TM this on a server that gets more than 40 peak players (Terry) ?

OverwatchVoice commented:
This isn't a bad idea on paper, but don't gas meters already do the same thing but more concisely and along with temperature information?
At first I assumed this shows visible gasses through the "windows", but the pump just red.
I can't see myself liking this change.

OverwatchVoice commented:
This isn't a bad idea on it's own, but gas meters already do the same thing but more concisely and along with temperature information?
At first I assumed this shows visible gasses through the "windows", at least giving information on gas within, but the pipe just turns red.
I can't see myself liking this change.
Also, the multi deck layer adapters have an error sprite overlaying them.

norsvenska commented:
multi-z pipes have an error on them

![image](https://github.com/tgstation/tgstation/assets/73006946/06ed0b38-445d-4752-83e8-2182185d1fa9)

(i'm assuming this pr is what's causing this)

ninjanomnom committed:
Fixes multiz pipe icon error

ninjanomnom committed:
Improves look of gas visual

ninjanomnom commented:
There's been a few error sprites popping up from previous pipe code using invalid icon states that was hidden by a lack of an error state, I've resolved the ones I've found so far. I've also improved the look a bit I think; I'm not much art inclined and this looks reasonable enough so I'll stick with what it is unless someone submits something better to use.

@rdragan 
>     * We already have gas meters which look much better and also work better at giving at-a-glance information
>     [...]
>     * Atmos already has major lag issues once pop goes over ~60 players and the atmos tech job becomes literally useless on highpop terry (consistently hitting 90 players every evening) as atmos crawls to a halt with 2 ticks per minute and even SM delams becoming impossible. Why are you adding additional load on a system that severely needs looking into more optimization?

Regarding the first, gas meters do not provide the same information, and I want it so you can't easily remove the display so that people who pass by the pipe can get some hints as to what is in the pipe.

As for the performance cost, as far as regular costs go this costs very nearly nothing, the only thing of possible concern is map tick from the vis contents which was supposedly much alleviated by backend byond changes a while ago.

LemonInTheDark commented on code/modules/atmospherics/machinery/pipes/pipes.dm:
```diff
@@ -73,25 +97,33 @@
 /obj/machinery/atmospherics/pipe/proc/update_pipe_icon()
-	icon = 'icons/obj/pipes_n_cables/pipes_bitmask.dmi'
+	switch(initialize_directions)
+		if(NORTH, EAST, SOUTH, WEST) // Pipes with only a single connection aren't handled by this system
```
what about multiz pipes?

ninjanomnom replied:
In short. the pipe icon file never had states for pipes with only a single connected dir. This is specifically to fix multi-z pipes as previously they ran this code but the icon they grabbed as a result was invalid. The icon file now has an error sprite in the blank icon state which meant that multi-z pipes were grabbing the error sprite.

Wallemations commented:
I stand by my previous statement that this should be something that isn't on all pipes by default.

Either a new windowed-pipe that's used in some places
Or
Using a screwdriver/other tool on a pipe can open up a little window after a longer do-after, perhaps?

I do like the idea of visually showing the gasses but there's some balance stuff that's a concern, and more importantly for me, I fear the windows will make all the pipes look too repetitious when they're all in the same pipeline.
By making them toggle-able we could spread them further apart, making them much nicer to look at

ninjanomnom commented:
> I stand by my previous statement that this should be something that isn't on all pipes by default.
> 
> Either a new windowed-pipe that's used in some places Or Using a screwdriver/other tool on a pipe can open up a little window after a longer do-after, perhaps?
> 
> I do like the idea of visually showing the gasses but there's some balance stuff that's a concern, and more importantly for me, I fear the windows will make all the pipes look too repetitious when they're all in the same pipeline. By making them toggle-able we could spread them further apart, making them much nicer to look at

I specifically don't want people to be able to remove the visual, if I were to do something like this I would force a window every other pipe or something like that. I don't feel the current design clutters things though and most people who have given me feedback on this have said they like the look.

lessthnthree commented:
They're using constant defined colors now, but I think that could be tweaked either by changing the shades of the defined colors (while keeping close enough to their container color, as the original defines are) or how they're mixed.

Distro air mix is dark grey-red, which is barely different than your empty pipe, for example.

![image](https://github.com/tgstation/tgstation/assets/83487515/95e94b2c-88b5-4323-8f36-94b076983a95)

Since nitrogen is going to be prevalent most everywhere, perhaps it should take a lower priority for determining the visible color. More often than not you're wanting to detect gases -other- than the presence of nitrogen, and if they're only causing a slight change to the nitrogen red this isn't the most useful.

ninjanomnom commented:
> Distro air mix is dark grey-red, which is barely different than your empty pipe, for example.

It's dark because there's not much gas in the pipe, the gif in the pr description has the same gas mixture in the second section of pipe but high enough pressure to reach max brightness.

> Since nitrogen is going to be prevalent most everywhere, perhaps it should take a lower priority for determining the visible color.

Giving gas types weights in the gas visual isn't a bad idea though and wouldn't be an issue adding, I'll experiment a bit with it

MrMelbert commented:
I don't know if this has been brought up yet but N2 looks too much like Plasma for what is a very harmless gas

![image](https://github.com/tgstation/tgstation/assets/51863163/074b21cb-7e8f-44be-9c19-e91a72b4f14e)

lessthnthree commented:
Not ideal when troubleshooting an SM For sure.

What if we changed N2 to be yellow?

We kind of have a sprite, it's not currently used by any other gas, and N2O being beige is still similar to yellow.

ninjanomnom commented:
I'm up for changing any colors though I feel I should also change the colors that tgui uses for gases as well if I do that since they're currently mirrored

lessthnthree commented:
Yeah that would be a one liner.

trerri commented:
yellow n2 would be great but would necessitate a yellow canister sprite

ninjanomnom commented:
We already have plenty gases with tgui colors that don't match their canisters, that part isn't a problem

ninjanomnom committed:
Changes nitrogen to yellow

ninjanomnom committed:
Adds corner gas visuals using the new low footprint style

Wallemations commented:
> I don't know if this has been brought up yet but N2 looks too much like Plasma for what is a very harmless gas
> 
> 
> 
> ![image](https://github.com/tgstation/tgstation/assets/51863163/074b21cb-7e8f-44be-9c19-e91a72b4f14e)
> 
> 

I always associated plasma with purple not red

MTandi commented:
This PR should be about the **Gas Meter** changing its color, not pipes having digital monitors displaying colors unrelated to the gas mix (like that N2 example and the red color).

The gas in pipes does not change frequently, so this adds more visual noise than usefulness. If you actually want to see this color, then you should plop a gas meter that will display this gas color.

And it would make more sense as the gas meter is already a digital device, so it's fine that it shows "digital" gasmix colors instead of the real ones.

I dislike this in its current state, because it makes maintenance look ugly for no good reason, and atmospherics department to have more visual noise.

OverwatchVoice commented:
The sprite for the display doesn't look like coder art, which I like, but...

> Changed N2 to yellow
This creates a mental inconsistency with already existing N2 canisters, which have always been marked **red**,
having N2 associated with two different colors. I'd suggest either having the N2 canisters also be **yellow**, or reverting both to **red** to clear that up.

lessthnthree commented:
> > Changed N2 to yellow
> 
> I disagree with [Wallemations](https://github.com/Wallemations) here on account of convention. Having N2 associated with two different colors creates a mental inconsistency with already existing N2 canisters, which have always been marked **red**. I'd suggest either having the N2 canisters also be **yellow** (which will inevitably create confusion from yellow being the default canister color being yellow already) or -- ideally -- reverting both to **red** to clear that up, since that is the way nitrogen always worked.
> 
> As for the core idea, I'm glad the sprites don't look like placeholder art anymore, but I'd still prefer the feature be rolled into gas meters.

Yellow with a white stripe instead of black hazard stripes works fine.

![image](https://github.com/tgstation/tgstation/assets/83487515/59bd5dd5-a34a-4f24-81c0-f042c8df935b)

Jacquerel commented on code/modules/atmospherics/machinery/datum_pipeline.dm:
```diff
@@ -280,3 +294,64 @@
 	for(var/datum/gas_mixture/gas_mixture as anything in gas_mixture_list)
 		gas_mixture.copy_from_ratio(total_gas_mixture, gas_mixture.volume / volume_sum)
+
+/////////////////////
```
```suggestion
```
Really minor nitpick but i think that big linesof `/` mess up our autodocs because it interprets the first three as starting an autodoc comment

Jacquerel commented on code/modules/atmospherics/machinery/pipes/pipe_spritesheet_helper.dm:
```diff
@@ -0,0 +1,148 @@
+		var/pending_icon_state = outputs[generated_icon]
+		generated_icons.Insert(generated_icon, pending_icon_state)
+
+/datum/pipe_icon_generator/proc/GeneratePipeDir(icon_state_suffix, layer, combined_dirs)
```
Is there a reason we're switching to pascal case for all these procs?

ninjanomnom replied:
That's how I've written procs when working on new code for a while now, and we don't have naming conventions. I only match nearby style if there is a style to match.

Jacquerel replied:
huh I guess that line in the style guide actually is only about typepaths, never mind then

LemonInTheDark replied:
oranges tried to make it cover procs too but burned out on it. so not a bad instinct

Jacquerel commented on code/modules/atmospherics/machinery/pipes/pipes.dm:
```diff
@@ -27,6 +34,23 @@
+		qdel(meter)
+	return ..()
+
+/////////////////
```
same here

Jacquerel reviewed (changes requested):
Just formatting nitpicks, this should be ready to go if it is done being testmerged?

LemonInTheDark commented:
Performance wise this is fine. it might have some slight clientside headaches but I doubt it's that bad.

I'm not 100% on how it looks, idk. I agree with wallem about clutter, it's for sure a thing in atmos, particularly with things like pipe bridges and 5 layers. more info and repetition makes me a bit worried

ninjanomnom commented:
I really don't feel it clutters much, especially with the much reduced visual size from how I had it initially. I could make it every other pipe or something but honestly it would probably be harder to read at a glance with inconsistent gaps.

ninjanomnom committed:
Dashes instead of slashes

SinfulBliss commented:
The yellow is pretty ugly.

ninjanomnom commented:
If people want to change colors they can feel free to in a pr, I don't care too much what colors people want to make them end up as.
